### DISCUSSION TITLE
EVP_CipherUpdate causing buffer overflow if olen is less then block size

### DISCUSSION THREAD

**Discussion by absrivas-in** [openssl/openssl#21610] [Q&A]
Hi Openssl Team, 
I am trying to encrypt the my input buffer using EVP_CipherUpdate api Openssl (3.0.8)
`int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl, const unsigned char *in, int inl)`
in which i have found that if we are passing outlen less then block size for example

Trying to use AES CBC mode with padding enabled where my input buffer size 14 bytes. 
but passing output buffer length as 2 ( consider it is as negative case).
Then EVP_EncryptUpdate should have thrown an error PROV_R_OUTPUT_BUFFER_TOO_SMALL but i am not getting any error 
API is returning 1 which is success return code. 

But it is causing buffer overflow and can lead to crash. 

Can you please help me to resolve this issue.

**Comment by paulidale** [comment]
The `outl` argument is output only.

As per the [manual page](https://www.openssl.org/docs/man3.0/man3/EVP_EncryptUpdate.html): _Encrypts inl bytes from the buffer in and writes the encrypted version to out. This function can be called multiple times to encrypt successive blocks of data. The amount of data written depends on the block alignment of the encrypted data. For most ciphers and modes, the amount of data written can be anything from **zero bytes to (inl + cipher_block_size - 1) bytes**_

Passing bad values to OpenSSL calls which then cause a crash is _not_ considered a bug in OpenSSL, it's a bug in your application code.

**Comment by absrivas-in** [comment]
Can you please give me some snippet to validate that outlen is sufficient or not to hold encrypted content.
Because:
e.g If i am passing my input buffer size of 14 bytes
then i am assuming that output length will be 16 bytes -> 14 Bytes + 2 bytes for padding for AES_CBC

But as per manual
inl + cipher_block_size - 1 -> 14 + 16 -1 = 19 bytes,
even though outlen is 16 bytes which is sufficient to hold encrypted content then why extra 3 bytes.

Is there any openssl api which can give expected outlen so that i can verify in application before calling EVP_CipherUpdate ?

**Reply by t8m** [reply]
Per manual it should be 29 bytes not 19 bytes. The EVP_CipherUpdate can output blocks from previous incomplete blocks.

Let's assume you're using ciphers that map 1 plaintext to 1 ciphertext byte plus eventual padding at the end (which is handled by EVP_CipherFinal() so that is not relevant for EVP_CipherUpdate). And let's assume the block size is 16 bytes.

You first call EVP_CipherUpdate with 15 bytes - that will not output anything because it is not a complete block.

Then you call EVP_CipherUpdate with another 17 bytes - the first byte will fill the first block and output it, the rest will fill another block and output it -> output size will be 2 blocks = 32 bytes. (17+16-1 = 32)

The application must either reserve one block plus or it must ensure it calls EVP_CipherUpdate with full blocks only (except for the last call which means the EVP_CipherFinal will output something).

**Comment by absrivas-in** [comment]
As you  have suggested i have done following condition check in my application before calling EVP_CipherUpdate :
I am returning  error for below cases
1) if olen < inlen
2) Get block size using api **EVP_CIPHER_CTX_get_block_size** 
    if olen <  inlen + cipher_block_size - 1

So in case of some Crypto mode like CFB  i am getting block size 1 and it is working fine. 

But for **AES CTS** crypto mode i am seeing some issue. 
AES_CTS is giving olen equivalent to  inLen after encryption. 
But  EVP_CIPHER_CTX_get_block_size is returning 16 bytes.
generally this behaviour is for blocksize 1 like ofb.

e.g: 
**Crypto mode :** AES CTS encryption
**Inlen:** 17 Byte
**olen:** 17 Byte

if olen <  inlen + cipher_block_size - 1  => 17 + 16 - 1 = 32 
17 < 32 hence returning the error.
But if you see for AES CTS mode olen 17 is valid buffer.

Please let me know what i am missing. 
Is above two condition is sufficient to check the olen have sufficient room and apart from AES CTS what other crypto mode will have same behaviour.

**Reply by absrivas-in** [reply]
Gentle reminder

**Comment by paulidale** [comment]
Did you read the manual page that was referenced above?  CTS modes have some specialnesess to is:

> - "cts" (**OSSL_CIPHER_PARAM_CTS**) <integer>
> 
> Gets 1 if the cipher algorithm _cipher_ uses ciphertext stealing,
> otherwise it gets 0.
> This is currently used to indicate that the cipher is a **one shot that only
> allows a single call to EVP_CipherUpdate(**).
> Use (EVP_CIPHER_get_flags(cipher) & EVP_CIPH_FLAG_CTS) to retrieve the
> cached value.

and later:
> - "cts_mode" (**OSSL_CIPHER_PARAM_CTS_MODE**) <UTF8 string>
> 
> Gets or sets the cipher text stealing mode. For all modes the output size is the
> same as the input size. The **input length must be greater than or equal to the
> block size**. (The block size for AES and CAMELLIA is 16 bytes).

**Comment by absrivas-in** [comment]
I think 
out buffer should be >=  inlen + cipher_block_size - 1 is not optimised approach in case of memory consumption.
for example AES CBC mode with padding 
output buffer length  for input buffer length from 16 to 31 will be 32 bytes

But based on the documentation as you have mentioned previously for blocksize = 16 and inputlen len 16 i observe one issue.

16+16-1 = 31  (inlen + cipher_block_size - 1) 
but outlen will be 32 byte which will cause buffer overflow 

now from the inlen from 16 to 31 based on the documentation second row value will be considered and returns the error from the application, 
But the actual outlen is third row which is optimised without any extra memory ask.  

```
input len:	 16	17	18	19	20	21	22	23	24	25	26	27	28	29	30	31	32
outputlen:	 31	32	33	34	35	36	37	38	39	40	41	42	43	44	45	46	47
actual outlen:   32	32	32	32	32	32	32	32	32	32	32	32	32	32	32	32	48

By further analysis i am seeing below mathematical expression to validate with received outlen and expected outlen.
for encryption:
expectedLength = ((inlen + blocksize - 1) / blocksize) * blocksize;

// To add extra block size for padding if inlen is multiple of block size
expectedLength += (isPad &&  (blocksize !=1) &&  (inlen % blocksize == 0)) ? blocksize : 0;

and for decryption:
expectedLen = (inlen/ blocksize) * blocksize;
```


I feel there should be an api in openssl to get the expected such length instead of application doing all this validation.
If it is already there then Please share so that i can integrate.

**Reply by t8m** [reply]
Adding such API would be non-trivial and it would require support for it to be added to all providers of ciphers including the third party ones. For that reason you would still have to handle the case when the proposed API call fails (because unsupported by underlying provider).

**Reply by absrivas-in** [reply]
So do we need to fix the documentation ? 
As i have mentioned that 
inlen + cipher_block_size - 1 is can cause the buffer overflow too.

**Reply by t8m** [reply]
But the padding happens in the EVP_CipherFinal when the last block is being output, isn't it so?

**Reply by absrivas-in** [reply]
Yes, you are right ! 
padding happens in  EVP_CipherFinal in which we are copying the final block to the  passed output buffer and that time we will face buffer overflow, 
hence i am suspecting inlen + cipher_block_size - 1 is not sufficient.

**Reply by t8m** [reply]
We were talking about EVP_CipherUpdate all the time not about EVP_CipherFinal in regards to the buffer overflow.

The documentation says: 

           If padding is enabled (the default) then EVP_EncryptFinal_ex()
           encrypts the "final" data, that is any data that remains in a
           partial block.  It uses standard block padding (aka PKCS padding)
           as described in the NOTES section, below. The encrypted final data
           is written to out which should have sufficient space for one cipher
           block. The number of bytes written is placed in outl. After this
           function is called the encryption operation is finished and no
           further calls to EVP_EncryptUpdate() should be made.

**Reply by t8m** [reply]
So you need one more full block buffer when you call EVP_CipherFinal (for encryption == EVP_EncryptFinal)

**Reply by absrivas-in** [reply]
I apologize for not mentioning the EVP_CipherFinal call earlier.

In summary, if padding is enabled, then before calling `EVP_EncryptUpdate`, we require a sufficient output buffer size of `inlen + blocksize - 1`, and an additional `blocksize` buffer size before calling `EVP_EncryptFinal`. This implies that a total buffer size of `inlen + 2 * blocksize - 1` is necessary. Have I understood this correctly?

Could you please inform me if you identify any issues in the validation excerpt provided below? I believe that the code snippet below accurately calculates the required output buffer size, accommodating block size 1 ciphers as well:

```c
blocksize = EVP_CIPHER_CTX_get_block_size(cipher_ctx);
expectedLength = ((inlen + blocksize - 1) / blocksize) * blocksize;

// Add an extra block size for padding if inlen is a multiple of block size
expectedLength += (isPad && (blocksize != 1) && (inlen % blocksize == 0)) ? blocksize : 0;
```

Is my understanding and analysis of the code snippet correct?

**Reply by paulidale** [reply]
You are correct about the sizes for the update and final call.  The documentation notes additional length for wrap ciphers, which you're not using here.

The total size will be less because padding is limited to one block, so you shouldn't need the extra block correction.  I.e. I don't think it will ever be the case that **both** the update and the final calls add a block.

**Reply by absrivas-in** [reply]
Hi @paulidale , 
Thank you for letting me know.

_"The total size will be less because padding is limited to one block, so you shouldn't need the extra block correction. I.e. I don't think it will ever be the case that both the update and the final calls add a block."_
i had the same impression but found that if padding is set and input length is multiple of block size which means (inlen % blocksize == 0) then final call will add extra 16 byte buffer. and i had added extra checks.

**Reply by paulidale** [reply]
That's because it isn't possible to tell if the final block has been padded or the encryption happened to result in something that looks like padding, so an extra block is needed in this case.

**Reply by absrivas-in** [reply]
Thank you @paulidale ,

Then below check is good i assume ?
```
blocksize = EVP_CIPHER_CTX_get_block_size(cipher_ctx);
expectedLength = ((inlen + blocksize - 1) / blocksize) * blocksize;

// Add an extra block size for padding if inlen is a multiple of block size
expectedLength += (isPad && (blocksize != 1) && (inlen % blocksize == 0)) ? blocksize : 0;
```

**Reply by paulidale** [reply]
Looks okay.

**Reply by absrivas-in** [reply]
Hi @paulidale 
Thank you for confirming that. 
May i know Is EVP_CIPHER_CTX_get_block_size is a reliable API to calculate expected output buffer length ?
My understanding is If cipher output length is equals to plaintext input length then it means that cipher mode is STREAM cipher and  
EVP_CIPHER_CTX_get_block_size should return 1 and EVP Final call will not add any padding bytes.
Is my understanding/assumption correct ?

**Reply by t8m** [reply]
Yes, if EVP_CIPHER_CTX_get_block_size() returns 1, it is a stream cipher and what you say applies.

**Reply by absrivas-in** [reply]
Thanks @t8m ,
I have a request that if below snippet looks good then can we please add this in Openssl EVP_CipherUpdate and EVP_CipherFinal_ex. 
```
blocksize = EVP_CIPHER_CTX_get_block_size(cipher_ctx);
expectedLength = ((inlen + blocksize - 1) / blocksize) * blocksize;

// Add an extra block size for padding if inlen is a multiple of block size
expectedLength += (isPad && (blocksize != 1) && (inlen % blocksize == 0)) ? blocksize : 0;
```

**Reply by t8m** [reply]
You mean in the manual pages for these functions?
